### DISCUSSION TITLE
Support for more modalities and base64 content

### DISCUSSION THREAD

**Discussion by marliessophie** [langfuse/langfuse#3004] [Ideas]
### Describe the feature or potential improvement

After our [latest release supporting images](https://langfuse.com/changelog/2024-08-21-openai-multi-modal-traces) (passed via url) in Langfuse, we're looking to support other modalities soon. 

We're committed to supporting **base64 images** in the near term. We're further planning support for **audio**, **video**, and **attachments** (e.g. PDFs). 

We'd love to gauge interest across the modalities to help with most efficient prioritisation.

**Comment by chai3** [comment]
+1 base64 image (most wanted)
+1 audio
+1 attachments(pdf, xlsx, html)

I dont need to support video in my use-case

**Comment by badbye** [comment]
Great! The dataset also need to support Multi-Modal input. Looking forward!

**Comment by Alpacolypse** [comment]
Eagerly awaiting base64 support, especially in markdown

**Comment by clemra** [comment]
A few double clicks on base64 as reactions to our most recent newsletter

**Comment by Hallicopter** [comment]
base64 images for sure.

**Comment by bioshazard** [comment]
+1 b64 images. Seems like the very next thing to support. Schema seems well established at this point.

**Reply by bioshazard** [reply]
Update: So I actually wound up writing a span intercept to strip b64 data and replace with filename. I plan on having a batch operation and worried it would balloon my DB. While looking at supporting other modalities, consider a way to decouple attachment reference so it can maybe resolve from a single source than to risk duplicating into the traces.

**Comment by tristanbonds** [comment]
Base64 images for multimodal models is the missing feature that is holding us back from using Langfuse. PDF attachments would also be very helpful.

**Reply by pleomax0730** [reply]
@tcbonds What alternative solution are you currently using? Our team is planning to add an AI observability tool, and your experience and insights would be greatly appreciated.

**Reply by tristanbonds** [reply]
Arize Phoenix. So far we have had a pretty good experience with it.

**Comment by profikid** [comment]
Bas64 or image url would be great to support

**Comment by wbreilly** [comment]
Audio is mission critical for us

**Comment by zxkxyz** [comment]
just here to add my support for base64 (because claude)

**Comment by marcklingen** [comment]
this is on the near-term roadmap, @hassiebp is working on it and specs look great, can't wait for the release!

**Comment by pleomax0730** [comment]
Here to support Audio, PDF source.

**Comment by darylsew** [comment]
+1 base64 (because claude)

**Comment by alexdanilowicz** [comment]
+1 base64 image (because claude)

**Comment by hassiebp** [comment]
Good news - we have just launched multimodal support in beta! 🎨 

- [Changelog post](https://langfuse.com/changelog)
- [Docs](https://langfuse.com/docs/tracing-features/multi-modality)
- [JS SDK release](https://github.com/langfuse/langfuse-js/releases/tag/v3.30.1)
- [Python SDK release](https://github.com/langfuse/langfuse-python/releases/tag/v2.54.1)

Please let us know your feedback here or create an issue if you encounter any bugs. Thanks! 🙏🏾

**Reply by marcklingen** [reply]
huge! looking forward to the feedback

**Reply by bioshazard** [reply]
DUDE YES! Y'all decoupled it to a bucket!! This is HUGE!! `LANGFUSE_S3_MEDIA_UPLOAD_`, now I can undo my customer span processor I was using to prevent the b64 from going into my DB.

**Comment by recursingfeynman** [comment]
vLLM multi-modal format, please! 

Here is an example from the [vLLM docs](https://docs.vllm.ai/en/v0.6.0/getting_started/examples/openai_audio_api_client.html):
```
{
    "type": "audio_url",
    "audio_url": {
        "url": audio_url
    },
}
```

It would also be great to add support for .ogg and .oga audios.

**Reply by marcklingen** [reply]
Thanks for sharing! I've created an issue for this. Open for contributions!
https://github.com/langfuse/langfuse/issues/4406

**Comment by pleomax0730** [comment]
I've encountered an issue. Please have a look.

https://github.com/orgs/langfuse/discussions/4363

**Reply by marcklingen** [reply]
thanks, already created an issue for this: https://github.com/langfuse/langfuse/issues/4410

**Comment by nitin302** [comment]
Does this support Anthropic models via AWS Bedrock using langchain or is there a better way?

`from langchain_aws import ChatBedrockConverse`

**Reply by marcklingen** [reply]
You can use Anthropic on Bedrock via
- Langfuse framework [integrations](https://langfuse.com/docs/integrations/overview): Langchain / LlamaIndex / AI SDK / ...
- Wrapping the Bedrock SDK, see example here: https://langfuse.com/docs/integrations/amazon-bedrock

**Reply by nitin302** [reply]
I was trying for multi-modal support (base64 images) - self hosted OSS, but was unable to get it working. Any working example please? Looking for both Amazon bedrock and Azure Open AI.

**Reply by marcklingen** [reply]
What kind of error do you get? Can you open an issue for this?

**Reply by nitin302** [reply]
I do not get any errors but images do not seem to render on the traces UI, so I am confused as to if I am running things the right way. If you can provide samples for Amazon bedrock and Azure Open AI, I can try this and re-confirm.

**Reply by marcklingen** [reply]
Can you share an example trace of your setup?

**Reply by nitin302** [reply]
@marcklingen i was able to get this working, it was an issue with my environment variable required for S3 upload. Quick question on the UI that renders the images... Is there any limits? For example, we have a single LLM call that send 10 images ... and we are not able to see it render inline, instead it shows up in the media section as attachments.

**Reply by nitin302** [reply]
however, with single images, it does render inline.

**Reply by marcklingen** [reply]
This might be the expected behavior. Can you share example traces?

cc @hassiebp

**Reply by nitin302** [reply]
you can just consider https://langfuse.com/guides/cookbook/example_multi_modal_traces#langchain-image-input but with multiple images...

**Reply by hassiebp** [reply]
@nitin302 Feel free to open a separate issue with a minimum repro example and langchain version here please: https://github.com/langfuse/langfuse/issues/new/choose

**Comment by mlcocdav** [comment]
Hi, I've tried the suggested ways to upload images using the API directly from https://langfuse.com/docs/tracing-features/multi-modality

But my images do not render correctly in the UI. I tried both Markdown with the image link and sending base64 encoded data URIs passing into the trace input field. 


This is the UI:
<img width="708" alt="image" src="https://github.com/user-attachments/assets/b443e876-21eb-4077-a0f3-490685d147f5" />
<img width="1056" alt="image" src="https://github.com/user-attachments/assets/4ff4a53f-551f-447a-b153-b66ea3810914" />

I want to see the images rendered.

Please help 🙏

**Comment by AviPeltz** [comment]
the ability to embed an iframe -> we would want to visualized 3d objects so either native support for a simple three.js renderer would be pretty easy to implement and nice, alternatively an iframe where we could add our own 3d renderer

**Comment by tom-diacono** [comment]
It would be really cool if this supported OTEL data. From from what I'm seeing so far it doesn't. If I could send my traces with the images in base64 (which I'm currently doing anyway), it would be great if Langfuse converted them to it's object storage, and have a link to them. Unfortunately for now, it's looking like I'm going to have to call an API to do this.

**Reply by marcklingen** [reply]
How do you integrate via OTel? A new major version of the sdks is in the works which will be otel native and do this transformation client side (direct s3 upload before otel traces are exported).

**Reply by tom-diacono** [reply]
Hi Marc,

Thanks for the reply, that's really good to know!  Both that the new versions will be OTEL Native, and that images will be taken into account! (Personally a big OTEL fan).

Our current implementation is using the Golang OTEL SDK, which as far as I can see, is a couple of versions behind the new OTEL semconv (meaning no event bodies, and slightly older `gen_ai` attributes, but seems to be working ok with Langfuse so far, I guess it still supports them).

So even when you guys get onto OTEL native, we'd probably have to roll our own implementation for images, given there's no Go SDK yet, and you're looking to handle images client side. I'm guessing we can still do that via API?


So far, though I may have mentioned this in a separate thread, I've not had much luck getting my implementation to work. I'm unsure whether it's just frontend rendering though. We've got multiple "chat" spans per trace, but our root span is not a "gen ai" one, it's a standard app trace. Is it the case that the frontend is only looking for images to render in the root span? I've noticed that in spans where I can see images (single call spans) I can only click on them when a top level spal is selected.

Can grab a screen grab tomorrow if you're interested

**Reply by marcklingen** [reply]
Can you share your current go implementation as a gist together with any errors that you get? Generally the media api returns presigned s3 upload links

**Reply by tom-diacono** [reply]
I've just re-read this and my messages are confusing even me, so first let me clarify:

* In the first message I was referring to hoping that I could send the full base64 encoded image data in the trace, and have Langfuse convert it for me (which, given you explanation, is going to be done client side with the new SDK's, so I'm guessing _probably_ won't end up being supported- all good, I think this one is answered)

* Second message is once I'd moved to doing the following:
    * Extract the image data from the traces client side
    * Send it to the Langfuse with the API call (which is successful)
    * Finally replace the image data in the trace (again, client side) with the Langfuse media token

So- what I'm trying to get to work is the images in the frontend. Below is an example trace of what it's currently looking like. You can see the media link, but there's no image link within the trace window.

![image](https://github.com/user-attachments/assets/69287bb3-2e95-4231-bcae-e2da85c01839)

I know the code works, as when I use it with a span of which the openai call is the root span, I can see the image.

![image](https://github.com/user-attachments/assets/cd929ce6-e90d-4e9d-85ee-6a2992ffd9ab)

The only thing I'm wondering- is will these images only show if the metadata containing them is in the root span, as opposed to being a child span? In the span in my second screenshot. I can only see the image when I'm focusing on the root one. If I focus on either child span, `openai_request` or `openai_completion`, there's again no image link. Given this last point, I'm wondering if this is by design?

**Reply by tom-diacono** [reply]
So to update- I finally got this to somewhat work.

The issue I was having was due to sending the wrong `Trace ID`. I was sending the `Span ID` instead :roll_eyes: .

Something it does highlight though (and I think I was alluding to this in my previous message, just not that I really knew it) is that it would be really cool if the media was stored per span that it was created in, as opposed to at the "Trace" level.

This would help as if there were many spans per trace, it would allow you to attribute the data to the correct span, as opposed to having all images at the top level, with no real order, or way to attribute it to a particular call

**Reply by git-hulk** [reply]
@tom-diacono You can have a try at https://github.com/git-hulk/langfuse-go, which supports most Langfuse APIs.

**Comment by jingtra** [comment]
Video would be great!

**Comment by barseghyanartur** [comment]
Please, add `DOCX` support.

**Comment by thomasdussaut** [comment]
As stated in this issue : https://github.com/langfuse/langfuse/issues/7319, in the meanwhile it would be a nice workaround to work with the dataset item POST endpoint. As for now, this route doesn't accept payloads bigger than 1MB. Increasing this limit to 10MB would allow to work with most PDF documents and images

**Comment by RuidongLiu** [comment]
need video support.

**Reply by Janscode** [reply]
same video support is probably a blocker for my organization

**Comment by Namir0806** [comment]
Please also support .flac file format for audio files. Thank you!

**Comment by Namir0806** [comment]
Please also support .ogg file format for audio files. Thank you!

**Reply by erickleal-azos** [reply]
WhatsApp's official audio format is .ogg. Since WhatsApp agents are highly relevant nowadays, it would be great if Langfuse supported this format.

**Comment by rileymiller** [comment]
csvs and xlsx would be great!!!

**Comment by marin1994** [comment]
html, markdown, but most important (as said above) video

**Comment by iyansr** [comment]
<img width="797" height="726" alt="image" src="https://github.com/user-attachments/assets/70b6029c-9b95-4cb7-9b1c-016ea5a31e3d" />
The base64 somehow not working (?)

**Reply by hassiebp** [reply]
Could you please create an issue with a trace link and repro example?

**Comment by iyansr** [comment]
@hassiebp I create issue here https://github.com/langfuse/langfuse/issues/10495

**Comment by JoostvDoorn** [comment]
For litellm it seems this works only when using the langfuse v2 integration as explained here https://docs.litellm.ai/docs/observability/langfuse_integration. With the v3 integration using  `litellm.callbacks = ["langfuse_otel"]` it will just store the base64 data and not display it in the interface.

**Comment by alexbensavanta** [comment]
Super tool out there! would love to see video rendering as well!
